### PR TITLE
SQL Pipelines Application Core Components

### PR DESCRIPTION
**Pull request by dli357**
Includes the specification generator, the job planner, and corresponding unit tests. Also fixed the projection transform to select fields in order.

### CODE REVIEW

File: sql-api/src/main/java/io/cdap/pipeline/sql/api/core/enums/OperandType.java [hunk 1]
```diff
@@ -16,11 +16,23 @@
...
-  CONSTANT, COLUMN
+  CONSTANT(Constant.class), COLUMN(Column.class);
+
+  private final Class operandClass;
+
+  OperandType(Class operandClass) {
+    this.operandClass = operandClass;
+  }
+
+  public Class getOperandClass() {
```
Comment (albertshau) [inline sql-api/src/main/java/io/cdap/pipeline/sql/api/core/enums/OperandType.java:35]: why is this needed?

File: sql-data-pipeline/pom.xml [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    </dependency>
+    <dependency>
+      <groupId>io.cdap.pipeline</groupId>
+      <artifactId>sql-plugins</artifactId>
+      <version>${project.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>io.cdap.cdap</groupId>
+      <artifactId>cdap-etl-proto</artifactId>
+      <version>6.1.0-SNAPSHOT</version>
```
Comment (albertshau) [inline sql-data-pipeline/pom.xml:106]: depend on non-snapshot versions unless you are using something in an unreleased version (don't think you are)

File: sql-data-pipeline/src/main/java/io/cdap/pipeline/sql/app/bigquery/BigQueryExecutor.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    GoogleCredentials credentials;
+    File credentialsPath = new File(serviceAccountPath);
+    FileInputStream serviceAccountStream = new FileInputStream(credentialsPath);
+    credentials = ServiceAccountCredentials.fromStream(serviceAccountStream);
+
+    // Instantiate a client.
+    BigQuery bigquery =
+      BigQueryOptions.newBuilder().setCredentials(credentials).build().getService();
+    QueryJobConfiguration queryConfig = QueryJobConfiguration.newBuilder(queries).build();
+    bigquery.query(queryConfig);
```
Comment (albertshau) [inline sql-data-pipeline/src/main/java/io/cdap/pipeline/sql/app/bigquery/BigQueryExecutor.java:117]: doesn't this return an operation or something like that? I was expecting there to be some polling logic along with some error handling

File: sql-data-pipeline/src/main/java/io/cdap/pipeline/sql/app/bigquery/BigQueryExecutor.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    for (StructuredStatement action: job.getActions()) {
+      if (action.getQuery() == null) {
+        throw new IllegalArgumentException("Unable to execute statement without a query.");
+      }
+      if (action.getTemporaryTable() != null) {
+        queryBuilder.append(serializer.getSQL(action.getQuery(), action.getTemporaryTable()));
+      } else if (action.getInsertIntoTable() != null) {
+        queryBuilder.append(serializer.getInsertIntoSQL(action.getQuery(), action.getInsertIntoTable()));
+      } else {
+        throw new IllegalArgumentException("Statement must contain a table declaration (for now).");
```
Comment (albertshau) [inline sql-data-pipeline/src/main/java/io/cdap/pipeline/sql/app/bigquery/BigQueryExecutor.java:102]: remove '(for now)'.

Also, these types of errors should cause things to fail early at app deployment, not later at runtime.

File: sql-data-pipeline/src/main/java/io/cdap/pipeline/sql/app/core/interfaces/SQLJob.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+package io.cdap.pipeline.sql.app.core.interfaces;
+
+import io.cdap.pipeline.sql.app.core.StructuredStatement;
+
+import java.util.List;
+
+/**
+ * Represents a SQL job to be executed.
+ */
+public interface SQLJob {
```
Comment (albertshau) [inline sql-data-pipeline/src/main/java/io/cdap/pipeline/sql/app/core/interfaces/SQLJob.java:26]: In general, only use an interface or abstract class if there is (or will soon be) more than one implementation. This doesn't seem to be the case here.

File: sql-data-pipeline/src/main/java/io/cdap/pipeline/sql/app/core/interfaces/SQLJob.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  List<StructuredStatement> getPreActions();
```
Comment (albertshau) [inline sql-data-pipeline/src/main/java/io/cdap/pipeline/sql/app/core/interfaces/SQLJob.java:27]: what is a pre action and post action? If everything needs to be run in order, why not just have a single list?

File: sql-data-pipeline/src/main/java/io/cdap/pipeline/sql/app/core/SQLPipelinePlanner.java [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+  public SQLJob plan(SQLPipelineSpec spec) {
+    // Simple validation
+    if (spec == null) {
+      throw new IllegalArgumentException("Input specification cannot be null");
+    }
+
+    // DAG traversal and query ordering logic
+    // Create query sets for each sink by traversing backwards from the sinks
+    // Add dependencies backwards so we don't use temporary tables before creating them
+    // This definition set is to prevent nodes from being defined multiple times
```
Comment (albertshau) [inline sql-data-pipeline/src/main/java/io/cdap/pipeline/sql/app/core/SQLPipelinePlanner.java:47]: I don't really understand the algorithm here but it seems backward. Generation of the queries should happen from the start to the end, in order to know the correct table names, properly validate schema, etc.

The Dag class already has a method to get the dag nodes back in topological order, you shouldn't need to write your own traversal logic.

**Review by albertshau — COMMENTED** [review]
there is a lot going on in this PR, with much of it likely changing if we switch to Calcite. In the future, try to break things up into smaller PRs that can be easily reviewed.

Can you shrink this down to only include the new module for the app along with skeleton classes for the app, workflow, and custom action?

There also seems to be several layers of abstraction and some serialization/deserialization that I don't think are adding much value, probably because it's following the same pattern done in the data-pipeline app. However, this app should be a good amount simpler, especially at this stage.

File: sql-api/src/main/java/io/cdap/pipeline/sql/api/core/enums/OperandType.java (hunk 1 above)
Comment (dli357) [reply sql-api/src/main/java/io/cdap/pipeline/sql/api/core/enums/OperandType.java:35]: Since the enums directly link the operand types to a particular class, it seemed cleaner to just get the class from the enum instead of creating switch statements everywhere. Is there a better pattern for handling this scenario?

File: sql-data-pipeline/pom.xml (hunk 2 above)
Comment (dli357) [reply sql-data-pipeline/pom.xml:106]: Fixed

File: sql-data-pipeline/src/main/java/io/cdap/pipeline/sql/app/core/interfaces/SQLJob.java (hunk 5 above)
Comment (dli357) [reply sql-data-pipeline/src/main/java/io/cdap/pipeline/sql/app/core/interfaces/SQLJob.java:26]: Fixed

File: sql-data-pipeline/src/main/java/io/cdap/pipeline/sql/app/core/interfaces/SQLJob.java (hunk 6 above)
Comment (dli357) [reply sql-data-pipeline/src/main/java/io/cdap/pipeline/sql/app/core/interfaces/SQLJob.java:27]: Originally, there was a guarantee that any pre-action would be executed before any action which must be executed before any post-action. This is necessary in cases where you have multiple sinks, and each sink has a corresponding job, but the executor must combine all jobs from sinks of the same kind (in this case, every sink runs on BigQuery so it must combine all the jobs into one job). At the moment we don't really have any particular cleanup/initialization actions that must be performed, but if we were to add view support it would require cleanup support at the very least.

File: sql-data-pipeline/src/main/java/io/cdap/pipeline/sql/app/bigquery/BigQueryExecutor.java (hunk 4 above)
Comment (dli357) [reply sql-data-pipeline/src/main/java/io/cdap/pipeline/sql/app/bigquery/BigQueryExecutor.java:102]: Fixed, and moved the check and exception into the configure method

File: sql-data-pipeline/src/main/java/io/cdap/pipeline/sql/app/core/SQLPipelinePlanner.java (hunk 7 above)
Comment (dli357) [reply sql-data-pipeline/src/main/java/io/cdap/pipeline/sql/app/core/SQLPipelinePlanner.java:47]: The table names and queries can be generated from just knowing the node and previous nodes which connect to it, so those do not require any kind of traversal to perform, but this might change once we add schema verification. The traversal is simply used for ensuring that the final ordering of the queries is in topological order.

I'll look into incorporating the existing DAG class since it already has logic to get the nodes in topological order.

File: sql-data-pipeline/src/main/java/io/cdap/pipeline/sql/app/bigquery/BigQueryExecutor.java (hunk 3 above)
Comment (dli357) [reply sql-data-pipeline/src/main/java/io/cdap/pipeline/sql/app/bigquery/BigQueryExecutor.java:117]: I believe that the query function throws errors with the API errors passed down if it happens. I was simply following this example:
https://cloud.google.com/bigquery/docs/reference/libraries#client-libraries-usage-java

There is a bit of a weird issue in the API where it throws a NullPointerException if the query fails under certain scenarios.

**Comment by dli357** [comment]
This PR is made obsolete by PR #5.
